pypowerwall/tedapi/ted_api_messages/BatteryComponentsMessage.py

```diff
@@ -0,0 +1,41 @@
+import logging
+import json
+from .. import tedapi_pb2
+from ..exceptions import PyPowerwallTEDAPIException
+from .TEDAPIMessage import TEDAPIMessage
+
+log = logging.getLogger(__name__)
+
+
+class BatteryComponentsMessage(TEDAPIMessage):
+    """ Protobuf message for requesting battery component data from the Powerwall. """
+
+    def __init__(self, din, pw_din):
+        super().__init__(din)
+        self.pw_din = pw_din
+
+    def get_message(self):
+        pb = tedapi_pb2.Message()
+        pb.message.deliveryChannel = 1
+        pb.message.sender.local = 1
+        pb.message.sender.din = self.din  # DIN of Primary Powerwall 3 / System
+        pb.message.recipient.din = self.pw_din  # DIN of Powerwall of Interest
+        pb.message.payload.send.num = 2
+        pb.message.payload.send.payload.value = 1
+        pb.message.payload.send.payload.text = " query ComponentsQuery (\n  $pchComponentsFilter: ComponentFilter,\n  $pchSignalNames: [String!],\n  $pwsComponentsFilter: ComponentFilter,\n  $pwsSignalNames: [String!],\n  $bmsComponentsFilter: ComponentFilter,\n  $bmsSignalNames: [String!],\n  $hvpComponentsFilter: ComponentFilter,\n  $hvpSignalNames: [String!],\n  $baggrComponentsFilter: ComponentFilter,\n  $baggrSignalNames: [String!],\n  ) {\n  # TODO STST-57686: Introduce GraphQL fragments to shorten\n  pw3Can {\n    firmwareUpdate {\n      isUpdating\n      progress {\n         updating\n         numSteps\n         currentStep\n         currentStepProgress\n         progress\n      }\n    }\n  }\n  components {\n    pws: components(filter: $pwsComponentsFilter) {\n      signals(names: $pwsSignalNames) {\n        name\n        value\n        textValue\n        boolValue\n        timestamp\n      }\n      activeAlerts {\n        name\n      }\n    }\n    pch: components(filter: $pchComponentsFilter) {\n      signals(names: $pchSignalNames) {\n        name\n        value\n        textValue\n        boolValue\n        timestamp\n      }\n      activeAlerts {\n        name\n      }\n    }\n    bms: components(filter: $bmsComponentsFilter) {\n      signals(names: $bmsSignalNames) {\n        name\n        value\n        textValue\n        boolValue\n        timestamp\n      }\n      activeAlerts {\n        name\n      }\n    }\n    hvp: components(filter: $hvpComponentsFilter) {\n      partNumber\n      serialNumber\n      signals(names: $hvpSignalNames) {\n        name\n        value\n        textValue\n        boolValue\n        timestamp\n      }\n      activeAlerts {\n        name\n      }\n    }\n    baggr: components(filter: $baggrComponentsFilter) {\n      signals(names: $baggrSignalNames) {\n        name\n        value\n        textValue\n        boolValue\n        timestamp\n      }\n      activeAlerts {\n        name\n      }\n    }\n  }\n}\n"
```

Comment:
General comment: Is it possible to break these out into more readable JSON so we can intelligently view what's truly in these queries?

Reply:
It doesn't actually seem like it is, if you even so much as remove that space before "query" it fails. I think the big binary chunk on the next line is a checksum very specific to that query.

Reply:
Ah, I learned from https://github.com/jasonacox/pypowerwall/pull/152 that you can in fact change some things about the query without triggering a checksum invalidate, but sad that it can't be done here.

Reply:
It looks like in that one you're changing the _arguments_ to the query, not the query itself, e.g. here you can see it accepts an argument called `msaComp` of some `ComponentFilterSmsaSignals` type:

![Screenshot 2025-03-23 at 15 18 56@2x](https://github.com/user-attachments/assets/992b1d16-0ad0-4115-8e7e-8034bec9d3e7)

So you can change the parameters you're passing, but the query itself is checksummed.

Reply:
I'm still surprised, given how locked down this generally is, that even that is allowed.

Reply:
Honestly I don't think it's a terrible approach for them, the alternatives are probably worse. Otherwise they'd have to be signing the arguments on the client side or building a more complicated API, and somebody's bound to break that anyway.

Reply:
I think the real reason is that this is utility-grade power equipment. It's a fact that's easy for us nerdy homeowners to forget, but they can't be letting every rando who thinks they have a "cool command" send arbitrary GraphQL to it. The attack surface area would be impossible to contain.

As much as I wish I could tweak the GraphQL to get parameters that I care about and/or remove ones I don't, I also understand why they do it.